src/developer-documentation/ibc/ibc.md

```diff
@@ -0,0 +1,237 @@
+    ///
+    /// Returns `NotAllowed` error if transfer is not approved.
+    ///
+    /// Returns `InsufficientBalance` error if `from` doesn't contain enough balance.
+    ///
+    /// Returns `SafeTransferCheckFailed` error if `to` doesn't accept transfer.
+    #[ink(message)]
+    fn transfer_from(
+        &mut self,
+        from: AccountId,
+        to: AccountId,
+        id: Id,
+        amount: Balance,
+        data: Vec<u8>,
+    ) -> Result<(), PSP37Error>;
+}
+```
+
+## 2.2 struct
+```js
+    pub struct Addr(String);
+
+    pub struct Coin {
+        pub denom: String,
+        pub amount: u128,
+    }
+
+    pub struct Cw20Coin {
+        pub address: String,
+        pub amount: u128,
+    }
+
+    pub struct TransferMsg {
+        /// The local channel to send the packets on
+        pub channel: String,
+        /// The remote address to send to.
+        /// Don't use HumanAddress as this will likely have a different Bech32 prefix than we use
+        /// and cannot be validated locally
+        pub remote_address: String,
+        /// How long the packet lives in seconds. If not specified, use default_timeout
+        pub timeout: Option<u64>,
+        /// An optional memo to add to the IBC transfer
+        pub memo: Option<String>,
+    }
+
+
+    pub enum Amount {
+        Native(Coin),
+        Cw20(Cw20Coin),
+    }
+
+    impl Amount {
+        pub fn from_parts(denom: String, amount: u128) -> Self {
+            if denom.starts_with("cw20:") {
+                let address = denom.get(5..).unwrap().into();
+                Amount::Cw20(Cw20Coin { address, amount })
+            } else {
+                Amount::Native(Coin { denom, amount })
+            }
+        }
+
+        pub fn cw20(amount: u128, addr: &str) -> Self {
+            Amount::Cw20(Cw20Coin {
+                address: addr.into(),
+                amount: amount,
+            })
+        }
+
+        pub fn native(amount: u128, denom: &str) -> Self {
+            Amount::Native(Coin {
+                denom: denom.to_string(),
+                amount: amount,
+            })
+        }
+    }
+
+    impl Amount {
+        pub fn denom(&self) -> String {
+            match self {
+                Amount::Native(c) => c.denom.clone(),
+                Amount::Cw20(c) => "cw20:".to_owned() + c.address.as_str(),
+            }
+        }
+
+        pub fn amount(&self) -> u128 {
+            match self {
+                Amount::Native(c) => c.amount,
+                Amount::Cw20(c) => c.amount,
+            }
+        }
+
+        /// convert the amount into u64
+        pub fn u64_amount(&self) -> Result<u64, Error> {
+            Ok(self.amount().try_into().unwrap())
+        }
+
+        pub fn is_empty(&self) -> bool {
+            match self {
+                Amount::Native(c) => c.amount == 0,
+                Amount::Cw20(c) => c.amount == 0,
+            }
+        }
+    }
+```
```

Comment:
```suggestion
```rust
pub struct Addr(String);

pub struct Coin {
    pub denom: String,
    pub amount: u128,
}

pub struct Cw20Coin {
    pub address: String,
    pub amount: u128,
}

pub struct TransferMsg {
    /// The local channel to send the packets on
    pub channel: String,
    /// The remote address to send to.
    /// Don't use HumanAddress as this will likely have a different Bech32 prefix than we use
    /// and cannot be validated locally
    pub remote_address: String,
    /// How long the packet lives in seconds. If not specified, use default_timeout
    pub timeout: Option<u64>,
    /// An optional memo to add to the IBC transfer
    pub memo: Option<String>,
}

pub enum Amount {
    Native(Coin),
    Cw20(Cw20Coin),
}

impl Amount {
    pub fn from_parts(denom: String, amount: u128) -> Self {
        if denom.starts_with("cw20:") {
            let address = denom.get(5..).unwrap().into();
            Amount::Cw20(Cw20Coin { address, amount })
        } else {
            Amount::Native(Coin { denom, amount })
        }
    }

    pub fn cw20(amount: u128, addr: &str) -> Self {
        Amount::Cw20(Cw20Coin {
            address: addr.into(),
            amount: amount,
        })
    }

    pub fn native(amount: u128, denom: &str) -> Self {
        Amount::Native(Coin {
            denom: denom.to_string(),
            amount: amount,
        })
    }
}

impl Amount {
    pub fn denom(&self) -> String {
        match self {
            Amount::Native(c) => c.denom.clone(),
            Amount::Cw20(c) => "cw20:".to_owned() + c.address.as_str(),
        }
    }

    pub fn amount(&self) -> u128 {
        match self {
            Amount::Native(c) => c.amount,
            Amount::Cw20(c) => c.amount,
        }
    }

    /// convert the amount into u64
    pub fn u64_amount(&self) -> Result<u64, Error> {
        Ok(self.amount().try_into().unwrap())
    }

    pub fn is_empty(&self) -> bool {
        match self {
            Amount::Native(c) => c.amount == 0,
            Amount::Cw20(c) => c.amount == 0,
        }
    }
}
```
```

Reply:
sure